module/zfs/zfs_ioctl.c

```diff
@@ -3258,22 +3311,30 @@ zfs_ioc_clone(const char *fsname, nvlist_t *innvl, nvlist_t *outnvl)
 {
 	int error = 0;
 	nvlist_t *nvprops = NULL;
+	nvlist_t *hidden_args = NULL;
+	dsl_crypto_params_t *dcp = NULL;
 	char *origin_name;
 
 	if (nvlist_lookup_string(innvl, "origin", &origin_name) != 0)
 		return (SET_ERROR(EINVAL));
 	(void) nvlist_lookup_nvlist(innvl, "props", &nvprops);
+	(void) nvlist_lookup_nvlist(innvl, ZPOOL_HIDDEN_ARGS, &hidden_args);
```

Comment:
> Do you think the atomicity of the crypto properties is more important enough to justify the special case code in dmu_objset_clone() and callees?

I think the "special case" code is better not necessarily because encryption is more important than other properties, but because it allows us to nicely share code. The `dmu_objset_create()` code does require the encryption properties atomically since encrypting the root directory of an encrypted filesystem will require the encryption key to be generated and stored on disk. So it at least requires the wrapping key and crypt. At that point, it makes sense to verify all the properties together. By creating clones the same way, we can share much of the code for creation and clones.